GafferSceneUI : Edit camera tooltips and descriptions

medubelko opened the pull request:
### Description
- Updates to the camera- and render option-related node descriptions and tooltips.
- Some of these tooltips were simply out of date, others are my attempts at improvements.
- Includes a single-sourcing hack that allows StandardOptions to inherit all its render option tooltips from Camera.
- Encoded one of the files in utf-8, which we should eventually do standard.

@johnhaddon There a a few cuts here and there, and I'm aware that you want to keep as much info as possible within the tooltips, but in-context I think my changes make more sense, and do a better job guiding the user as they work through the plugs. YMMV.

### Checklist ###

- [x] I have read the [contribution guidelines](https://github.com/GafferHQ/gaffer/blob/master/CONTRIBUTING.md).
- [x] I have updated the documentation, if applicable.
- [x] I have tested my change(s) in the test suite, and added new test cases where necessary.
- [x] My code follows the Gaffer project's prevailing coding style and conventions.

medubelko commented on python/GafferSceneUI/CameraUI.py:
```diff
@@ -46,59 +49,237 @@
+		aperture and the image are different. The following fitting
+		modes are available:
+
+		<ul>
```
We lose some of the source readability here, but this list markup works in both the app and the Reference docs, which are a higher priority for me.

johnhaddon replied:
Ugh. I have a branch where I do the markdown->html conversion on the fly for tooltips. Let me see if I can dust that off so we can keep the source readable.

johnhaddon replied:
I've made a PR for markdown-in-tooltips here : https://github.com/GafferHQ/gaffer/pull/3065. If you could take a look that would be great...

medubelko commented on python/GafferSceneUI/CameraUI.py:
```diff
@@ -46,59 +49,237 @@
 # Metadata
 ##########################################################################
 
+__overridesMetadata = {}
```
This to L270 should probably be looked at by someone more experienced in Python than I.

medubelko commented:
Rebased with a reformat using Markdown. Now you can review the content itself.

medubelko commented on python/GafferSceneUI/CameraUI.py:
```diff
@@ -46,51 +46,207 @@
 # Metadata
 ##########################################################################
 
+__overridesMetadata = {}
```
This to L249 should probably be looked at by someone other than me.

medubelko commented:
John, can I merge this, or did you not finish your review?

medubelko commented:
@themissingcow @johnhaddon Could I get an update on this one?

themissingcow commented on python/GafferSceneUI/StandardOptionsUI.py:
```diff
@@ -106,6 +106,8 @@ def __statisticsSummary( plug ) :
 	scene. These should be respected by all renderers.
 	""",
 
+	## For some of these tooltips, we hackily grab the metadata from
```
We shouldn't be 'hackily' doing anything here.

The pattern in `CameraTweaksUI` that interrogates the metadata system can be used, or if we are sharing the dict instead, we shouldn't be relying on any order-dependence. You simply `import GafferSceneUI.CameraUI` at the top if you want to make use of it. This works even whilst the module is importing as you're actually just directly asking it to load the child module. If you tried to use `import GafferSceneUI` and access it from there that would then become order dependent.

themissingcow replied:
Checked with @johnhaddon, we should use the shared dict approach over interrogating metadata....

themissingcow commented on python/GafferSceneUI/CameraUI.py:
```diff
@@ -46,59 +49,234 @@
+	__overridesMetadata[overrideKeyName] = overrideMetadata
+
+
+## Utility hack to allow appending two dicts within the scope of an
```
See below, but we should get rid of this.

medubelko commented on python/GafferSceneUI/CameraUI.py:
```diff
@@ -1,3 +1,6 @@
+#!/usr/local/bin/python
+# -*- coding: utf-8 -*-
```
When I discussed this with John, the biggest concern was whether the unicode could survive the round-trip to QT and the docs (it does). With the _Python Editor_ capable of handling unicode, are there other parts of the API or code that can read tooltips to make this an issue?

johnhaddon replied:
I think this is fine - the [source encoding](https://www.python.org/dev/peps/pep-0263/) feature is available from Python 2.3 onwards, and we require 2.7. In Python 3, the default encoding is UTF8 anyway, and we can remove the tag.

medubelko commented on python/GafferSceneUI/CameraUI.py:
```diff
@@ -152,8 +338,7 @@
 			"""
-			The horizontal field of view, specified in degrees.  When using a perspective projection
-			with "Field Of View" control mode, this will control the focal length.
+			The horizontal field of view, in degrees.
```
I removed the extra detail because it was irrelevant to users. To see that plug, _Field of View_ mode has to already be active.

Now, from a reference perspective, you're right. I was doing my best to shorten all tooltips for the users.

themissingcow replied:
I guess this just wasn't the longest tooltip to start with?

johnhaddon replied:
> I removed the extra detail because it was irrelevant to users.

Let's be clear about something : the description metadata is not just used as a tooltip, it is also used for the reference documentation. We shouldn't be "improving" the former at the expense of the latter. I don't think that one extra piece of information in an already tiny description is hurting, and "the users" use the reference documentation too.

medubelko commented on python/GafferSceneUI/CameraUI.py:
```diff
@@ -178,40 +367,33 @@
-			of the aperture.
-
-			Note that in CG applications, "Film Back" and aperture are used interchangeably - lenses
-			are treated as perfectly symmetric boxes, with the aperture at the front the same size
```
I think I know which bits you're talking about, but in my opinion neither the tips nor the reference docs are where you should be learning first-hand about Gaffer's camera model. What effect values have? Sure. But for this particular plug, a discussion about how Gaffer treats lenses takes up precious space. That's what _Anatomy of a Camera_ is for.

To me, at least.

medubelko commented on python/GafferSceneUI/CameraUI.py:
```diff
@@ -178,40 +367,33 @@
-			camera settings, but also the resolution and film fit mode in the render settings.
+			The width and height of the aperture. Use this in
+			conjunction with the focal length to set the camera's
+			projection. It can be a measure in any unit of length, as
```
I chose _projection_ for a few reasons:

I was thinking that users familiar with photography would prefer to think in terms of focal length, not field of view or angle of view. Maybe I'm coloured by personal experience – when I dabbled in amateur photography, I don't remember FoV being a frequent term.

There's also the potential to muddy things by saying that the result of aperture is a FoV. As you mentioned above, one of the salient points is that Gaffer always stores projection as `aperture` and `focalLength` parameters. So, the result, as it were, of `aperture` should be the specific projection of the camera.

Speaking of it more generally as projection also leaves the door open to show the correlation to `aperture` in orthographic cameras, where field of view isn't relevant.

themissingcow replied:
I guess I was thinking more along the lines that there are 3 related quantities we surface in the UI that affect the resultant frustum for a perspective projection:  Aperture, Focal Lenth and FoV. You can only change two of those at any time as one is always a product of the other two.

Aside from any more abstract meaning of those terms in the broader topic of photography - it seemed that in the original text (for this and other tool tips in this section), the use of FoV in this was about highlighting that interaction. By switching to projection does it loosen that link as it is it is anchoring against a fourth term, instead of one of the other, previously referenced properties? 

Perhaps more importantly though, is that we already have a first-class 'projection' plug on the Camera node? Stating:

>  use this in conjunction with the focal length to set the camera's projection.

Could be confusing for the reader as the 'Projection' plug (persp/ortho) is entirely unaffected by any changes here?

johnhaddon replied:
I'm with Tom here I'm afraid : the original text was more comprehensive and more precise. I'm concerned that in an attempt to make things "friendlier" we're actually just making them vaguer and less complete. I think articles and tutorials are the place to be less formal and more hand-holdy, but the metadata needs to be about cold hard facts, precisely communicated.

medubelko commented on python/GafferSceneUI/CameraUI.py:
```diff
@@ -249,25 +438,34 @@
 			"""
-			Setting a non-zero fStop will enable focal blur in renderers that support it.
-
-			fStop specifies ratio of focal length divided by lens aperture.
```
I honestly can't remember. I'll restore it.

medubelko commented on python/GafferSceneUI/CameraUI.py:
```diff
@@ -276,17 +474,25 @@
+			be (1,1). To compensate, select _Custom_ and then input a
+			value that scales the scene unit of measure to a realistic
+			aperture size. For example, 0.35 would convert 1 centimeter
+			(Alembic/USD default) to 35mm, which would simulate a 35mm
```
The last paragraph was meant more as a tip, because _Field of View_ mode uses an aperture of 1x1mm, which isn't obvious unless you look at the camera's parameters in the inspector.

themissingcow replied:
Sorry, my comment wasn't so much about having a 'tip', more about switching to talking about focal length rather than aperture. @danieldresser do you have any thoughts?

medubelko commented:
Gah, GitHub confusion. Ignore my "review." You can see my comments in the threads started by your review.

themissingcow commented on python/GafferSceneUI/CameraUI.py:
```diff
@@ -46,59 +49,234 @@
 import GafferSceneUI
 
-filmFitMetadata = [
+renderOptionsMetadata = {
```
Just checked with @johnhaddon  - we actually prefer the method you're using here (ie: not `MetadataAlso`), but we shoudl still move it to `StandardOptionsUI.py`

themissingcow commented on python/GafferSceneUI/CameraUI.py:
```diff
@@ -276,17 +474,25 @@
+
+			For example, take a 50mm focal length, and a scene that
+			uses decimeters (dm) for its world space units. The
+			mm-to-dm conversion would be: `50mm × 0.01 = 0.05dm`.
```
Sorry had missed this before. I'm not sure this is a helpful example, as the 50mm confuses things. Isn't it just a case of "If you've used mm for f/l and your scene is in decimetres, pick 'mm to decimetres' in the dropdown?" No-one needs to do a calculation unless they're in FOV mode and need accurate bokeh?

medubelko commented:
@themissingcow @danieldresser-ie After a losing a bit more hair over this, I think I've finally addressed all of your comments with edits you should both be happy with. Please see the GDoc I shared and give them a look over. Thanks.

danieldresser-ie commented:
Yep, I'm happy

medubelko commented:
Addressed all the Python issues and updated the text.

themissingcow commented on python/GafferSceneUI/StandardOptionsUI.py:
```diff
@@ -96,281 +97,422 @@ def __statisticsSummary( plug ) :
+
+}
+
+plugsMetadata.update( renderOptionsMetadata )
```
Whats happening here?  The plugs on the nodes seem to be under `options.*` do we need to copy all the entires to non-prefixed plug names?

medubelko replied:
Addressed in 1c2a4237a4d47008ec2014616029106b956c7b1c.

themissingcow commented on python/GafferSceneUI/StandardOptionsUI.py:
```diff
@@ -96,281 +97,422 @@ def __statisticsSummary( plug ) :
 	return ", ".join( info )
 
-Gaffer.Metadata.registerNode(
+renderOptionsMetadata = {
```
Don't know if I'm missing somthing, but do we need to split the metadata up like this (between `renderOptionsMetadata` and `plugMetadata`)? Is it just because they have the `.options` prefix on this node? Seems fragile and confuses what the source of truth is and why it's separate? Took me 10 mins to work out what the separation was reading from cold. Could just put it all in one?

The code in other classes still works, it just becomes `renderOptionsMetadata["options.filmFit"]` instead of `["filmFit"]`.  The transformation in `CameraUI` becomes:
`overrideKeyName = keyName.replace( "options.", "renderSettingOverrides." )` + a little logic to find the `descriptions` entry, afaict.

themissingcow replied:
Digesting some more - if you're concerned about having to filter out layout sections etc... then maybe at least make this just a simple descriptions dictionary and put _all_ the descriptions in. So its `{ <name> : <description }`, and have `renderOptionsPresets` for `filmFit` which seems to be the only one that shares anything other than a description.  It's annoying we're suck with the old python bindings that mean you have paired values in an array.

We just don't want to have `StandardOptions` metadata being split up in to parts just because `CameraUI` wants to access some of the data. That might change in the future and StandardOptions shouldn't have to be updated if say, we decide to expose `motionBlur` as an override.... so if we're making the concession for sharing, then at least make sure we're consistent within this file about how we handle metadata... That split of why some plugs are in there and some aren't take a bit of reverse engineering to understand.

medubelko replied:
Fair points. Partly I was just following what you suggested, and partly I'm just not fully aware of all the methods available for manipulating Python dicts. I'll improve it.

medubelko replied:
Addressed in 1c2a4237a4d47008ec2014616029106b956c7b1c.

medubelko commented on python/GafferSceneUI/CameraTweaksUI.py:
```diff
@@ -57,10 +57,7 @@ def __populateMetadata():
-				Arbitrary numbers of user defined tweaks may be
-				added as children of this plug via the user
-				interface, or using the CameraTweaks API via python.
+				Add a camera tweak.
```
That's a good point. It would be more consistent with some of the other reversions we did.

medubelko replied:
Addressed in 87eb0ddf027605becb838cc64174076983829ed2.

medubelko commented:
So I don't have to rebase prematurely again, let's make sure my refactor passes muster.

themissingcow commented on python/GafferSceneUI/CameraUI.py:
```diff
@@ -327,140 +321,66 @@
+			option = plainKey,
+			description = overrideMetadata[1]
+		)
+		__overridesMetadata[overrideKey] = overrideMetadata
```
If you wanted to make this a bit more readable, you could simplify this (from line 358) to:

```
__overridesMetadata[overrideKey] = [
    "description",
    "Overrides the `{option}` render option:\n\n{description}".format(
        option = plainKey,
        description = value[1]
    )
]

medubelko replied:
eb0d6dcb98f7300fcde77d24def32010037d5d54

themissingcow commented on python/GafferSceneUI/StandardOptionsUI.py:
```diff
@@ -96,281 +97,330 @@ def __statisticsSummary( plug ) :
 	return ", ".join( info )
 
-Gaffer.Metadata.registerNode(
+plugsMetadata = {
```
This is much easier to understand now, thanks. It might be worth a quick comment here along the lines of:
```
# This dict is shared with other nodes that hold related plugs. It is 
# assumed that the first metadata element for each plug is always
# `description`, so take care to preserve this when updating entires here.
```

medubelko replied:
eb0d6dcb98f7300fcde77d24def32010037d5d54

themissingcow commented:
@medubelko We do have test failures though now that need fixing.

medubelko commented:
> We do have test failures though now that need fixing before a merge
Forgot about the doc algo test. Bah!

Anyhow, fixed a larger issue with the plug names being different between nodes. Not sure if you'll find my solution Pythonic enough.

themissingcow commented:
> Anyhow, fixed a larger issue with the plug names being different between nodes. 

Ahh nice one.

> Not sure if you'll find my solution Pythonic enough.

Just to make a little easier to read, how about something along the lines of the following. The dictionary lookup avoids the double loop of iterating all override name for each plug. If you're going to be changing the names of some of them too, we might as well include all the name changes directly in the mapping dictionary:

```
# Maps plugs from the StandardOptionsUI to our override plugs
__plugsToOverride = {
	"options.filmFit" : "renderSettingOverrides.filmFit",
	"options.renderResolution":  "renderSettingOverrides.resolution",
	...
]

for srcPlug, overridePlug in __plugsToOverride.items() :

    plugMetadata = GafferSceneUI.StandardOptionsUI.plugsMetadata[ srcPlug ]

     __overridesMetadata[ overridePlug ] = [
        "description",
        "Overrides the `{option}` render option:\n\n{description}".format(
            option = srcPlugName.replace( "options.", ""),
            # We assume the second element is the description
            description = plugMetadata[1]
        )
    ]
```

medubelko commented:
Addressed in 2db42d7481dfb6ff02d614dae9166bd8663350a6. There was an extra bit of getting the `filmFit` presets working properly. Hopefully this is good now. If it is, I can rebase and merge tomorrow.
